### PR TITLE
Extend porous navier stokes

### PR DESCRIPTION
**Pull request by jgonzalezusua**
**📝 Description**
This PR is aimed to include the porous version of Navier-Stokes via VMS for both projections, ASGS and OSS.

Please mark the PR with appropriate tags: 
- Applications
- Feature

**🆕 Changelog**
- Renovated `alternative_qs_vms_dem_coupled.cpp`, `alternative_d_vms_dem_coupled.cpp`, `qs_vms_dem_coupled.cpp` and `d_vms_dem_coupled.cpp`
- Added support for quadratic elements
- Included tests for quadratic elements to test second derivatives of the shape functions
- Added porosity term in cpp tests of the `alternative_qs_vms_dem_coupled.cpp`, `alternative_d_vms_dem_coupled.cpp`, `qs_vms_dem_coupled.cpp` and `d_vms_dem_coupled.cpp` elements.

### CODE REVIEW

File: applications/FluidDynamicsApplication/custom_elements/fluid_element.h
Comment (rubenzorrilla) [inline applications/FluidDynamicsApplication/custom_elements/fluid_element.h]: Why is this required to be `virtual`?

File: applications/FluidDynamicsApplication/custom_strategies/schemes/bdf2_turbulent_scheme.h
Comment (rubenzorrilla) [inline applications/FluidDynamicsApplication/custom_strategies/schemes/bdf2_turbulent_scheme.h]: Why is the `virtual` required?

File: applications/FluidDynamicsApplication/custom_strategies/schemes/residualbased_simple_steady_scheme.h
Comment (rubenzorrilla) [inline applications/FluidDynamicsApplication/custom_strategies/schemes/residualbased_simple_steady_scheme.h]: This is doing nothing. Please remove it and leave current base class implementation.

File: applications/FluidDynamicsApplication/tests/cpp_tests/test_alternative_qs_vms_dem_coupled_quadratic_element.cpp
Comment (rubenzorrilla) [inline applications/FluidDynamicsApplication/tests/cpp_tests/test_alternative_qs_vms_dem_coupled_quadratic_element.cpp]: I'd add this test to current `test_alternative_qs_vms_dem_coupled_element.cpp`. Note that you're testing the very same implementation but with different template instantiation.

File: applications/FluidDynamicsApplication/tests/cpp_tests/test_alternative_qs_vms_dem_coupled_quadratic_element_3D.cpp
Comment (rubenzorrilla) [inline applications/FluidDynamicsApplication/tests/cpp_tests/test_alternative_qs_vms_dem_coupled_quadratic_element_3D.cpp]: I'd add this test to current `test_alternative_qs_vms_dem_coupled_element.cpp` under the very same reasoning I posted above.

File: applications/FluidDynamicsApplication/custom_strategies/schemes/bdf2_turbulent_scheme.h
Comment (jgonzalezusua) [reply applications/FluidDynamicsApplication/custom_strategies/schemes/bdf2_turbulent_scheme.h]: Because this function is derived in `bdf2_turbulent_schemeDEMCoupled.h`, which will be merge to master after this PR.

File: applications/FluidDynamicsApplication/tests/cpp_tests/test_alternative_qs_vms_dem_coupled_quadratic_element_3D.cpp
Comment (jgonzalezusua) [reply applications/FluidDynamicsApplication/tests/cpp_tests/test_alternative_qs_vms_dem_coupled_quadratic_element_3D.cpp]: I thought that it would be better to split these tests in three in order to ease the detection of errors. However, I can add them in only one.

File: applications/FluidDynamicsApplication/custom_strategies/schemes/residualbased_simple_steady_scheme.h
Comment (jgonzalezusua) [reply applications/FluidDynamicsApplication/custom_strategies/schemes/residualbased_simple_steady_scheme.h]: Yes, it was for a testing implementation, I forgot to remove it.

File: applications/FluidDynamicsApplication/custom_elements/fluid_element.h
Comment (jgonzalezusua) [reply applications/FluidDynamicsApplication/custom_elements/fluid_element.h]: This function is derived in the elements I have renovated. I need to initialize the second derivatives of the shape functions.

File: applications/FluidDynamicsApplication/custom_strategies/schemes/bdf2_turbulent_scheme.h
Comment (rubenzorrilla) [reply applications/FluidDynamicsApplication/custom_strategies/schemes/bdf2_turbulent_scheme.h]: But this is already `virtual` in the base class... I mean, it makes no sense to add it also here.

File: applications/FluidDynamicsApplication/tests/cpp_tests/test_alternative_qs_vms_dem_coupled_quadratic_element_3D.cpp
Comment (rubenzorrilla) [reply applications/FluidDynamicsApplication/tests/cpp_tests/test_alternative_qs_vms_dem_coupled_quadratic_element_3D.cpp]: The tests are already split by the macro.

File: applications/FluidDynamicsApplication/custom_elements/fluid_element.h
Comment (rubenzorrilla) [reply applications/FluidDynamicsApplication/custom_elements/fluid_element.h]: The point is that by doing so you're introducing the vtable overhead. Considering that you know at compile time which elements are high order as we've a dimension and number of nodes template argument, I think we can achieve this in a more generic and efficient way by using an `if constexpr` together with an overload of the method that calculates the kinematics. Note that this approach would make possible to keep a unique implementation for all the elements deriving from the `FluidElement`.

File: applications/FluidDynamicsApplication/custom_strategies/schemes/bdf2_turbulent_scheme.h
Comment (jgonzalezusua) [reply applications/FluidDynamicsApplication/custom_strategies/schemes/bdf2_turbulent_scheme.h]: It is true

File: applications/FluidDynamicsApplication/tests/cpp_tests/test_alternative_qs_vms_dem_coupled_quadratic_element_3D.cpp
Comment (jgonzalezusua) [reply applications/FluidDynamicsApplication/tests/cpp_tests/test_alternative_qs_vms_dem_coupled_quadratic_element_3D.cpp]: Ok, so you mean to write them in the same .cpp file. I thought you meant to join them in only one test.  To me it is ok

Comment (rubenzorrilla) [reply applications/FluidDynamicsApplication/tests/cpp_tests/test_alternative_qs_vms_dem_coupled_quadratic_element_3D.cpp]: Exactly this.

File: applications/FluidDynamicsApplication/custom_elements/fluid_element.h
Comment (jgonzalezusua) [reply applications/FluidDynamicsApplication/custom_elements/fluid_element.h]: Ok, but we shall consider to implement `GetShapeSecondDerivatives` in `FluidElement` and modify 'CalculateLocalVelocityContribution' using an `if constexpr`.

Comment (jgonzalezusua) [reply applications/FluidDynamicsApplication/custom_elements/fluid_element.h]: Do you agree with that, @rubenzorrilla?

Comment (rubenzorrilla) [reply applications/FluidDynamicsApplication/custom_elements/fluid_element.h]: No. `GetShapeSecondDerivatives` must be removed. Note that all the kinematics are computed at the `CalculateGeometryData`. Hence, my suggestion is to do overload of the `CalculateGeometryData` so we have two signatures, current one, without second order derivatives and a new one, with first and second order ones for high order geometries. Then you can filter which one is used at compile time by using and `if constexpr` with the dimension and number of nodes template argument (indeed, we can make a static member var to filter this).

Pointing that these changes must be definitively done in a separate PR.

Comment (jgonzalezusua) [reply applications/FluidDynamicsApplication/custom_elements/fluid_element.h]: @rubenzorrilla, I am still confused. If I remove `GetShapeSecondDerivatives`, where do I implement the transform of the second shape derivatives?

Comment (rubenzorrilla) [reply applications/FluidDynamicsApplication/custom_elements/fluid_element.h]: In a new overload of the `CalculateGeometryData`.

Comment (jgonzalezusua) [reply applications/FluidDynamicsApplication/custom_elements/fluid_element.h]: @rubenzorrilla, this means that we must implement `GetShapeSecondDerivatives` in the `fluid_element` and overload the `UpdateIntegrationPointData` to call the second derivatives from data.

Comment (rubenzorrilla) [reply applications/FluidDynamicsApplication/custom_elements/fluid_element.h]: No, you do another version of the `CalculateGeometryData` (method overload) that does what we're currently doing plus the second derivatives.

**Comment by RiccardoRossi** [comment]
is this done?

**Comment by jgonzalezusua** [comment]
> is this done?

@RiccardoRossi, in theory it is done, but we have decided to move the second derivatives transform to KratosCore. We are waiting for the approval of #11253.

**Comment by jgonzalezusua** [comment]
It is very strange since in my local machine all tests run OK. But in github, the results in fluid cpp tests are different... How is it possible?

**Comment by rubenzorrilla** [comment]
> It is very strange since in my local machine all tests run OK. But in github, the results in fluid cpp tests are different... How is it possible?

It could be a memory error. Are you running the tests in FullDebug?

**Comment by jgonzalezusua** [comment]
> > It is very strange since in my local machine all tests run OK. But in github, the results in fluid cpp tests are different... How is it possible?
> 
> It could be a memory error. Are you running the tests in FullDebug?

Yes, I am running in FullDebug. That's why is so strange.

**Comment by jgonzalezusua** [comment]
@rubenzorrilla tests are runnning. We are ready to merge.

File: applications/FluidDynamicsApplication/custom_elements/fluid_element.h
Comment (rubenzorrilla) [reply applications/FluidDynamicsApplication/custom_elements/fluid_element.h]: It seems that this is not solved yet.

**Review by rubenzorrilla — COMMENTED** [review]
The issue with the `virtual` seems to be still pending. Aside of this everything looks good to me.

**Comment by jgonzalezusua** [comment]
@rubenzorrilla, I have removed the `virtual` from `fluid_element.h`

File: applications/FluidDynamicsApplication/custom_elements/alternative_d_vms_dem_coupled.cpp
Comment (rubenzorrilla) [inline applications/FluidDynamicsApplication/custom_elements/alternative_d_vms_dem_coupled.cpp:191]: As these are template parameters, in this case you can use `if constexpr` so you evaluate the if at compile time.

(I'd do this in a future PR considering that this is already taking a while...)

**Review by rubenzorrilla — APPROVED** [review]
No more changes in base elements. Thanks.

File: applications/FluidDynamicsApplication/custom_elements/alternative_d_vms_dem_coupled.cpp
Comment (jgonzalezusua) [reply applications/FluidDynamicsApplication/custom_elements/alternative_d_vms_dem_coupled.cpp:191]: Yeah, you are right

**Pull request merged**
